File: monitor.py
import asyncio
import json
import os
import urllib.request
import websockets
from dotenv import load_dotenv
from notifier import notify_status_change

load_dotenv()

DISCORD_TOKEN = os.getenv("DISCORD_TOKEN")
CHANNEL_ID = os.getenv("CHANNEL_ID")
GATEWAY_URL = "wss://gateway.discord.gg/?v=9&encoding=json"
API_BASE = "https://discord.com/api/v9"
POLL_INTERVAL = 1.5  # Poll every 1.5 seconds (safe under rate limit)

last_channel_name = None


def get_current_channel_name():
    """Fetch the current channel name via REST API."""
    url = f"{API_BASE}/channels/{CHANNEL_ID}"
    req = urllib.request.Request(url)
    req.add_header("Authorization", DISCORD_TOKEN)
    req.add_header("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) Chrome/120.0.0.0")
    try:
        with urllib.request.urlopen(req, timeout=5) as response:
            data = json.loads(response.read().decode())
            return data.get("name")
    except Exception as e:
        print(f"Poll error: {e}")
        return None


async def poll_channel():
    """Backup polling loop - checks channel name every POLL_INTERVAL seconds."""
    global last_channel_name
    while True:
        await asyncio.sleep(POLL_INTERVAL)
        name = get_current_channel_name()
        if name and last_channel_name and name != last_channel_name:
            print(f"[POLL] Name changed from '{last_channel_name}' to '{name}'")
            notify_status_change(name)
            last_channel_name = name
        elif name and not last_channel_name:
            last_channel_name = name


async def send_identify(ws):
    identify_payload = {"op": 2, "d": {"token": DISCORD_TOKEN, "properties": {"$os": "linux", "$browser": "Chrome", "$device": "Chrome"}}}
    await ws.send(json.dumps(identify_payload))
    print("Sent Identify payload")


async def heartbeat(ws, interval):
    while True:
        await asyncio.sleep(interval / 1000)
        await ws.send(json.dumps({"op": 1, "d": None}))
        print("Sent heartbeat")


async def listen():
    global last_channel_name

    async with websockets.connect(GATEWAY_URL, max_size=10_000_000) as ws:
        # Receive Hello (op 10)
        hello = json.loads(await ws.recv())
        heartbeat_interval = hello["d"]["heartbeat_interval"]
        print(f"Connected. Heartbeat interval: {heartbeat_interval}ms")

        # Start heartbeat task
        asyncio.create_task(heartbeat(ws, heartbeat_interval))

        # Send Identify
        await send_identify(ws)

        # Listen for events
        async for message in ws:
            data = json.loads(message)
            op = data.get("op")
            t = data.get("t")

            # Handle heartbeat ACK
            if op == 11:
                print("Heartbeat ACK received")
                continue

            # Handle CHANNEL_UPDATE
            if op == 0 and t == "CHANNEL_UPDATE":
                channel_data = data.get("d", {})
                channel_id = channel_data.get("id")
                channel_name = channel_data.get("name")

                if channel_id == CHANNEL_ID:
                    print(f"Channel update detected: {channel_name}")
                    if last_channel_name is not None and channel_name != last_channel_name:
                        print(f"Name changed from '{last_channel_name}' to '{channel_name}'")
                        notify_status_change(channel_name)
                    last_channel_name = channel_name


async def run_both():
    """Run both WebSocket listener and polling concurrently."""
    await asyncio.gather(listen(), poll_channel())


def main():
    global last_channel_name

    if not DISCORD_TOKEN:
        print("Error: DISCORD_TOKEN not set in .env")
        return
    if not CHANNEL_ID:
        print("Error: CHANNEL_ID not set in .env")
        return

    # Fetch and display current channel name
    last_channel_name = get_current_channel_name()
    if last_channel_name:
        print(f"Current channel name: '{last_channel_name}'")
    else:
        print("Warning: Could not fetch initial channel name")

    print(f"Monitoring channel: {CHANNEL_ID}")
    print(f"Polling every {POLL_INTERVAL}s + WebSocket real-time events")
    asyncio.run(run_both())


if __name__ == "__main__":
    main()

File: src/monitor.rs
//! Dual-mode Discord channel monitoring via REST polling and WebSocket Gateway.
//!
//! This module provides two concurrent monitoring strategies:
//! - REST polling: Periodically fetches channel info via Discord API
//! - WebSocket: Real-time updates via Discord Gateway

use crate::models::{Channel, GatewayMessage, HelloPayload, IdentifyPayload, IdentifyProperties};
use crate::notifier::Notifier;
use futures_util::{SinkExt, StreamExt};
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::RwLock;
use tokio_tungstenite::{connect_async, tungstenite::Message};

const DISCORD_API_BASE: &str = "https://discord.com/api/v9";
const DISCORD_GATEWAY_URL: &str = "wss://gateway.discord.gg/?v=9&encoding=json";
const USER_AGENT: &str = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";

/// Check for channel name changes and notify if changed.
///
/// This helper extracts the common pattern used in both poll_loop and websocket_loop
/// to avoid code duplication.
async fn check_and_notify_change(
    new_name: Option<String>,
    last_name: &Arc<RwLock<Option<String>>>,
    notifier: &Arc<Notifier>,
    source: &str,
) {
    let last = last_name.read().await;
    if *last != new_name {
        drop(last);
        let mut last_write = last_name.write().await;
        *last_write = new_name.clone();
        drop(last_write);
        if let Some(ref name) = new_name {
            println!("[{}] Channel name changed to: {}", source, name);
            notifier.start_alarm(name).await;
        }
    }
}

/// Fetch channel name from Discord REST API.
///
/// Returns `Ok(Some(name))` if the channel exists and has a name,
/// `Ok(None)` if the channel exists but has no name (e.g., DM channels),
/// or an error if the request fails.
pub async fn fetch_channel_name(
    token: &str,
    channel_id: &str,
) -> Result<Option<String>, reqwest::Error> {
    let client = reqwest::Client::new();
    let url = format!("{}/channels/{}", DISCORD_API_BASE, channel_id);

    let response = client
        .get(&url)
        .header("Authorization", token)
        .header("User-Agent", USER_AGENT)
        .send()
        .await?
        .error_for_status()?;

    let channel: Channel = response.json().await?;
    Ok(channel.name)
}

/// Poll Discord REST API for channel name changes.
///
/// This loop runs indefinitely, checking for channel name changes
/// at the specified interval. When a change is detected, it triggers
/// the notifier alarm.
pub async fn poll_loop(
    token: String,
    channel_id: String,
    poll_interval: f64,
    notifier: Arc<Notifier>,
    last_name: Arc<RwLock<Option<String>>>,
) {
    let interval = Duration::from_secs_f64(poll_interval);

    loop {
        tokio::time::sleep(interval).await;

        match fetch_channel_name(&token, &channel_id).await {
            Ok(current_name) => {
                check_and_notify_change(current_name, &last_name, &notifier, "POLL").await;
            }
            Err(e) => {
                eprintln!("[POLL] Failed to fetch channel: {}", e);
            }
        }
    }
}

/// Connect to Discord Gateway and listen for CHANNEL_UPDATE events.
///
/// This function:
/// 1. Connects to the Discord WebSocket Gateway
/// 2. Handles the Hello message and extracts heartbeat interval
/// 3. Sends Identify payload with browser spoofing
/// 4. Spawns a heartbeat task
/// 5. Listens for CHANNEL_UPDATE events and triggers alarms on changes
pub async fn websocket_loop(
    token: String,
    channel_id: String,
    notifier: Arc<Notifier>,
    last_name: Arc<RwLock<Option<String>>>,
) {
    loop {
        println!("[WS] Connecting to Discord Gateway...");

        match connect_async(DISCORD_GATEWAY_URL).await {
            Ok((ws_stream, _)) => {
                println!("[WS] Connected to Gateway");

                let (mut write, mut read) = ws_stream.split();

                // Wait for Hello message (op 10)
                let heartbeat_interval = match read.next().await {
                    Some(Ok(Message::Text(text))) => {
                        match serde_json::from_str::<GatewayMessage>(&text) {
                            Ok(msg) if msg.op == 10 => {
                                if let Some(d) = msg.d {
                                    match serde_json::from_value::<HelloPayload>(d) {
                                        Ok(hello) => {
                                            println!(
                                                "[WS] Received Hello, heartbeat_interval: {}ms",
                                                hello.heartbeat_interval
                                            );
                                            hello.heartbeat_interval
                                        }
                                        Err(e) => {
                                            eprintln!("[WS] Failed to parse Hello payload: {}", e);
                                            continue;
                                        }
                                    }
                                } else {
                                    eprintln!("[WS] Hello message missing 'd' field");
                                    continue;
                                }
                            }
                            Ok(msg) => {
                                eprintln!("[WS] Expected op 10, got op {}", msg.op);
                                continue;
                            }
                            Err(e) => {
                                eprintln!("[WS] Failed to parse Gateway message: {}", e);
                                continue;
                            }
                        }
                    }
                    Some(Ok(_)) => {
                        eprintln!("[WS] Expected text message for Hello");
                        continue;
                    }
                    Some(Err(e)) => {
                        eprintln!("[WS] WebSocket error: {}", e);
                        continue;
                    }
                    None => {
                        eprintln!("[WS] Connection closed before Hello");
                        continue;
                    }
                };

                // Send Identify (op 2)
                let identify = GatewayMessage {
                    op: 2,
                    t: None,
                    d: Some(
                        serde_json::to_value(IdentifyPayload {
                            token: token.clone(),
                            properties: IdentifyProperties {
                                os: "linux".to_string(),
                                browser: "Chrome".to_string(),
                                device: "Chrome".to_string(),
                            },
                        })
                        .expect("Failed to serialize identify properties"),
                    ),
                };

                let identify_json = serde_json::to_string(&identify)
                    .expect("Failed to serialize identify payload");
                if let Err(e) = write.send(Message::Text(identify_json)).await {
                    eprintln!("[WS] Failed to send Identify: {}", e);
                    continue;
                }
                println!("[WS] Sent Identify payload");

                // Spawn heartbeat task
                let heartbeat_interval_ms = heartbeat_interval;
                let (heartbeat_tx, mut heartbeat_rx) = tokio::sync::mpsc::channel::<()>(1);

                let heartbeat_handle = tokio::spawn(async move {
                    let interval = Duration::from_millis(heartbeat_interval_ms);
                    loop {
                        tokio::time::sleep(interval).await;
                        if heartbeat_tx.send(()).await.is_err() {
                            break;
                        }
                    }
                });

                // Main event loop
                let channel_id_clone = channel_id.clone();
                let notifier_clone = Arc::clone(&notifier);
                let last_name_clone = Arc::clone(&last_name);

                loop {
                    tokio::select! {
                        // Handle heartbeat
                        Some(()) = heartbeat_rx.recv() => {
                            let heartbeat = GatewayMessage {
                                op: 1,
                                t: None,
                                d: None,
                            };
                            let heartbeat_json = serde_json::to_string(&heartbeat)
                                .expect("Failed to serialize heartbeat payload");
                            if let Err(e) = write.send(Message::Text(heartbeat_json)).await {
                                eprintln!("[WS] Failed to send heartbeat: {}", e);
                                break;
                            }
                        }

                        // Handle incoming messages
                        msg = read.next() => {
                            match msg {
                                Some(Ok(Message::Text(text))) => {
                                    if let Ok(gateway_msg) = serde_json::from_str::<GatewayMessage>(&text) {
                                        // Handle CHANNEL_UPDATE (op 0, t: "CHANNEL_UPDATE")
                                        if gateway_msg.op == 0 {
                                            if let Some(ref t) = gateway_msg.t {
                                                if t == "CHANNEL_UPDATE" {
                                                    if let Some(d) = gateway_msg.d {
                                                        if let Ok(channel) = serde_json::from_value::<Channel>(d) {
                                                            if channel.id == channel_id_clone {
                                                                check_and_notify_change(
                                                                    channel.name,
                                                                    &last_name_clone,
                                                                    &notifier_clone,
                                                                    "WS",
                                                                ).await;
                                                            }
                                                        }
                                                    }
                                                }
                                            }
                                        }
                                        // Handle heartbeat ACK (op 11) - just acknowledge
                                        else if gateway_msg.op == 11 {
                                            // Heartbeat acknowledged, continue
                                        }
                                    }
                                }
                                Some(Ok(Message::Close(_))) => {
                                    println!("[WS] Connection closed by server");
                                    break;
                                }
                                Some(Err(e)) => {
                                    eprintln!("[WS] WebSocket error: {}", e);
                                    break;
                                }
                                None => {
                                    println!("[WS] Connection closed");
                                    break;
                                }
                                _ => {}
                            }
                        }
                    }
                }

                // Clean up heartbeat task
                heartbeat_handle.abort();
            }
            Err(e) => {
                eprintln!("[WS] Failed to connect: {}", e);
            }
        }

        // Wait before reconnecting
        println!("[WS] Reconnecting in 5 seconds...");
        tokio::time::sleep(Duration::from_secs(5)).await;
    }
}

/// Run the complete dual-mode monitoring system.
///
/// This function:
/// 1. Fetches the initial channel name
/// 2. Runs both polling and WebSocket loops concurrently
pub async fn run_monitor(token: String, channel_id: String, sound_path: String) {
    let notifier = Arc::new(Notifier::new(sound_path));
    let last_name: Arc<RwLock<Option<String>>> = Arc::new(RwLock::new(None));

    // Fetch initial channel name
    println!("Fetching initial channel state...");
    match fetch_channel_name(&token, &channel_id).await {
        Ok(name) => {
            println!("Initial channel name: {:?}", name);
            let mut last = last_name.write().await;
            *last = name;
        }
        Err(e) => {
            eprintln!("Failed to fetch initial channel state: {}", e);
        }
    }

    // Run both monitoring modes concurrently
    let poll_token = token.clone();
    let poll_channel_id = channel_id.clone();
    let poll_notifier = Arc::clone(&notifier);
    let poll_last_name = Arc::clone(&last_name);

    let ws_token = token;
    let ws_channel_id = channel_id;
    let ws_notifier = Arc::clone(&notifier);
    let ws_last_name = Arc::clone(&last_name);

    println!("Starting dual-mode monitoring (REST polling + WebSocket)...");

    tokio::join!(
        poll_loop(poll_token, poll_channel_id, 1.5, poll_notifier, poll_last_name),
        websocket_loop(ws_token, ws_channel_id, ws_notifier, ws_last_name)
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_constants() {
        assert!(DISCORD_API_BASE.starts_with("https://"));
        assert!(DISCORD_GATEWAY_URL.starts_with("wss://"));
        assert!(USER_AGENT.contains("Mozilla"));
    }

    #[test]
    fn test_api_url_construction() {
        let channel_id = "123456789";
        let url = format!("{}/channels/{}", DISCORD_API_BASE, channel_id);
        assert_eq!(url, "https://discord.com/api/v9/channels/123456789");
    }

    #[tokio::test]
    async fn test_last_name_rwlock_behavior() {
        let last_name: Arc<RwLock<Option<String>>> = Arc::new(RwLock::new(None));

        // Initial state
        {
            let read = last_name.read().await;
            assert!(read.is_none());
        }

        // Write new value
        {
            let mut write = last_name.write().await;
            *write = Some("test-channel".to_string());
        }

        // Read new value
        {
            let read = last_name.read().await;
            assert_eq!(*read, Some("test-channel".to_string()));
        }
    }
}
